test/throttler/remote_throttler.js

```diff
@@ -0,0 +1,171 @@
+      'operation is not set so operation should not be allowed'
+    );
+    throttler._credits[operation] = 0;
+    assert.isNotOk(throttler._isAllowed(operation), 'operation is set but lacks credit');
+  });
+
+  it("return false for isAllowed if operation doesn't have enough credits", () => {
+    throttler._credits[operation] = 0.5;
+    assert.isNotOk(throttler._isAllowed(operation));
+  });
+
+  it('succeed when we retrieve credits for multiple operations', done => {
+    throttler.setProcess({ uuid: uuid });
+    server.addCredits(serviceName, [
+      { operation: operation, credits: 5 },
+      { operation: other_operation, credits: 3 },
+    ]);
+    throttler._credits[operation] = 0;
+    throttler._credits[other_operation] = 0;
+    throttler._onCreditsUpdate = _throttler => {
+      assert.isOk(_throttler.isAllowed(operation));
+      assert.equal(_throttler._credits[operation], 4);
+      assert.isOk(_throttler.isAllowed(other_operation));
+      assert.equal(_throttler._credits[other_operation], 2);
+      assert.equal(LocalBackend.counterValue(metrics.throttlerUpdateSuccess), 1);
+      done();
+    };
+    throttler._refreshCredits();
+  });
+
+  it('emit failure metric on failing to query for credits', done => {
+    throttler.setProcess({ uuid: uuid });
+    throttler._credits[operation] = 0;
+    metrics.throttlerUpdateFailure.increment = function() {
+      assert.equal(logger._errorMsgs.length, 1, `errors=${logger._errorMsgs}`);
+      done();
+    };
+    throttler._host = 'Llanfair­pwllgwyngyll­gogery­chwyrn­drobwll­llan­tysilio­gogo­goch';
+    throttler._refreshCredits();
+  });
+
+  it('emit failure metric on failing to parse bad http json response', done => {
+    throttler.setProcess({ uuid: uuid });
+    throttler._credits[operation] = 0;
+    metrics.throttlerUpdateFailure.increment = function() {
+      assert.equal(logger._errorMsgs.length, 1, `errors=${logger._errorMsgs}`);
+      done();
+    };
+    server.addCredits(serviceName, 'not-json');
+    throttler._refreshCredits();
+  });
+
+  it('emit failure metric when server returns an invalid response', done => {
+    throttler.setProcess({ uuid: uuid });
+    throttler._credits[operation] = 0;
+    metrics.throttlerUpdateFailure.increment = function() {
+      assert.equal(logger._errorMsgs.length, 1, `errors=${logger._errorMsgs}`);
+      done();
+    };
+    throttler._refreshCredits();
+  });
+
+  it('not fetch credits if no operations have been seen', () => {
+    throttler = new RemoteThrottler(serviceName);
+    throttler.setProcess({ uuid: uuid });
+    throttler._refreshCredits();
+    assert.equal(Object.keys(throttler._credits).length, 0);
+  });
+
+  it('refresh credits after _afterInitialDelay is called', done => {
+    throttler.setProcess({ uuid: uuid });
+    throttler._credits[operation] = 0;
+    server.addCredits(serviceName, [{ operation: operation, credits: 5 }]);
+    throttler._onCreditsUpdate = _throttler => {
```

Comment:
Looks like you can avoid overwriting the private(-ish) property by using the `options.onCreditUpdate`. This would avoid relying on the private property and breaking the public API would break the test.

```
let creditUpdatedHook;

beforeEach(() => {
  creditUpdatedHook = sinon.spy();
  throttler = new RemoteThrottler(serviceName, {
    // ...
    onCreditUpdate: (...args) => creditUpdatedHook(...args),
  });
});

it('....', () => {
  assertTrue(creditUpdatedHook.called);
});

it('...', done => {
  creditUpdatedHook = () => {
    // ...
    done();
  }
});
```